### PR TITLE
vcsim: ResourcePool Create/Destroy support

### PR DESCRIPTION
**Pull request by dougm**

### CODE REVIEW

File: pkg/vsphere/simulator/cluster_compute_resource.go [hunk 1]
```diff
@@ -81,8 +81,8 @@ func CreateClusterComputeResource(f *Folder, name string, spec types.ClusterConf
 	cluster := &ClusterComputeResource{}
 	cluster.Name = name
 
-	pool := esx.ResourcePool
-	Map.PutEntity(cluster, Map.NewEntity(&pool))
+	pool := NewResourcePool()
```
Comment (hickeng) [inline pkg/vsphere/simulator/cluster_compute_resource.go]: Does this map to /datacenter/host/cluster/**Resources** ?

File: pkg/vsphere/simulator/property_collector.go [hunk 2]
```diff
@@ -40,6 +41,37 @@ func NewPropertyCollector(ref types.ManagedObjectReference) object.Reference {
...
+func getObject(ref types.ManagedObjectReference) (reflect.Value, bool) {
+	obj := Map.Get(ref)
+	if obj == nil {
+		return reflect.Value{}, false
+	}
+
+	rval := reflect.ValueOf(obj).Elem()
+	rtype := rval.Type()
+
+	// PropertyCollector is for Managed Object types only (package mo).
```
Comment (hickeng) [inline pkg/vsphere/simulator/property_collector.go]: Is the emulating the behaviour we saw with ObjectName and networks?

File: pkg/vsphere/simulator/property_collector.go [hunk 3]
```diff
@@ -40,6 +41,37 @@ func NewPropertyCollector(ref types.ManagedObjectReference) object.Reference {
...
+	// We need to dig out the mo type for PropSet.All to work properly and
+	// for the case where the type has a field of the same name, for example:
+	// mo.ResourcePool.ResourcePool
+	for {
+		if path.Base(rtype.PkgPath()) != "mo" {
+			if rtype.Kind() != reflect.Struct || rtype.NumField() == 0 {
+				log.Printf("%#v does not have an embedded mo type", ref)
+				return reflect.Value{}, false
+			}
+			rval = rval.Field(0)
```
Comment (hickeng) [inline pkg/vsphere/simulator/property_collector.go]: Is this going to work for vApp? I know it "extends" a ResourcePool but I'm unsure how that translates to the vmodl structures.

File: pkg/vsphere/simulator/cluster_compute_resource.go (hunk 1 above)
Comment (dougm) [inline pkg/vsphere/simulator/cluster_compute_resource.go]: The root pool is already there by default:
```
% govc ls -l host/DC0_C0
/DC0/host/DC0_C0/DC0_C0_H0 (HostSystem)
/DC0/host/DC0_C0/DC0_C0_H1 (HostSystem)
/DC0/host/DC0_C0/DC0_C0_H2 (HostSystem)
/DC0/host/DC0_C0/Resources (ResourcePool)
```
This supports creating child pools:
```
% govc ls -l host/DC0_C0/Resources/*

% govc pool.create DC0_C0/Resources/child

% govc ls -l host/DC0_C0/Resources/*
/DC0/host/DC0_C0/Resources/child (ResourcePool)

% govc ls -l host/DC0_C0/Resources/child/*

% govc pool.create DC0_C0/Resources/child/grandkid

% govc ls -l host/DC0_C0/Resources/child/*
/DC0/host/DC0_C0/Resources/child/grandkid (ResourcePool)
```

File: pkg/vsphere/simulator/property_collector.go (hunk 2 above)
Comment (dougm) [inline pkg/vsphere/simulator/property_collector.go]: We had already been behaving the same way as vSphere in that respect, by virtue of how Go's reflection works.

File: pkg/vsphere/simulator/property_collector.go (hunk 3 above)
Comment (dougm) [inline pkg/vsphere/simulator/property_collector.go]: Yes, it already works in similar cases such as with `ClusterComputeResource` which extends `ComputeResource`.

**Comment by emlin** [comment]
Will this work for VM creation under one resource pool, and then query back that VM through "vm" property of one resource pool?
Currently, I cannot query back any VMs from the root resource pool.

**Comment by dougm** [comment]
@emlin yes, this line adds the reference(s) you need to do that: https://github.com/vmware/vic/pull/1149/files#diff-c237d870267fcc0ac773daaa9ad1c637R199

**Comment by caglar10ur** [comment]
LGTM :sheep: :it:

**Comment by hickeng** [comment]
LGTM

**Comment by dougm** [comment]
♻️

**Pull request merged**